[Frontend] Add `-exit-on-first-error`

hamishknight opened the pull request:
This terminates the compiler as soon as an error is encountered.

rdar://123630771

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

xedin commented on lib/AST/DiagnosticEngine.cpp:
```diff
@@ -1521,6 +1522,11 @@ void DiagnosticEngine::emitDiagnostic(const Diagnostic &diagnostic) {
+  // If we've been told to exit on the first error, we're done if we have
+  // an error.
+  if (getExitOnFirstError() && info && info->Kind == DiagnosticKind::Error)
+    exit(EXIT_FAILURE);
```
This is a bit more fatal than I'd expect... I wonder if in some situations we'd end up existing without producing any diagnostics because consumers didn't have time to render them?

DougGregor replied:
I don't think we should have `exit` anywhere in the compiler. Can we put in a reasonable set of "bail out" points at various places in the compiler where we just stop doing more work (e.g., stop walking declarations in TypeCheckDeclPrimary, stop looping through function declarations, etc.)? Then we can have more interesting heuristics, such as bailing out from disastrous compiles after the first N errors .

xedin replied:
I agree, that’s more inline what I had in mind as well.

hamishknight replied:
> I wonder if in some situations we'd end up existing without producing any diagnostics because consumers didn't have time to render them?

Ah yeah good point we should definitely be calling `flush` & `finishProcessing` on them, beyond that I don't _think_ there can be an issue there?

> Can we put in a reasonable set of "bail out" points at various places in the compiler where we just stop doing more work (e.g., stop walking declarations in TypeCheckDeclPrimary, stop looping through function declarations, etc.)?

My worry with this is that it will need to be done in a lot of places, and even then we could still run into cases where the compiler kicks off a bunch of work we don't want. My feeling is that this serves 2 main purposes, first as a debugging aid for compiler engineers, similar to `-swift-diagnostics-assert-on-error`, secondly as a workaround for developers that can be used in cases where the compiler raises an error quickly but then proceeds to spin for a while. For the former, `exit` definitely seems like the right thing. For the latter, I'm not _too_ confident we could cover every possible case where the compiler could potentially spin.

> Then we can have more interesting heuristics, such as bailing out from disastrous compiles after the first N errors 

Is there any reason we can't have that with `exit`?

> I don't think we should have `exit` anywhere in the compiler

I should note that we have a couple of uses already:

https://github.com/swiftlang/swift/blob/b4b99e9d28232b7ae2d37534b9fa2bf91162e490/lib/Serialization/DeserializeSIL.cpp#L439

https://github.com/swiftlang/swift/blob/b4b99e9d28232b7ae2d37534b9fa2bf91162e490/lib/SIL/Verifier/SILVerifier.cpp#L145

https://github.com/swiftlang/swift/blob/b4b99e9d28232b7ae2d37534b9fa2bf91162e490/lib/SILOptimizer/PassManager/PassManager.cpp#L430

DougGregor replied:
> > Then we can have more interesting heuristics, such as bailing out from disastrous compiles after the first N errors
> 
> Is there any reason we can't have that with `exit`?

The problem with `exit` is that it takes down the whole process, and the whole process can be consist of more than this one compilation job. It could be a batch job where other translation units should run, a SourceKit process, some future compile server, etc. 

> > I don't think we should have `exit` anywhere in the compiler
> 
> I should note that we have a couple of uses already

Yes, and I don't think we should have these. The pass manager one, for example, should clearly produce an error but keep going---there is no reason to exit there. The DeserializeSIL one replaced a crash/assertion, and recovery is hard, but it shouldn't exit because you lose useful information (e.g., serialized diagnostics) and take down the whole process.

hborla replied:
>For the latter, I'm not too confident we could cover every possible case where the compiler could potentially spin.

I think the primary subsystem that can spin on invalid code, because invalid code fundamentally increases the complexity, is the constraint system. I think an initial version of this flag which simply doesn't advance to the "next" subsystem after a previous one produces failures would address the main pain points of that second use case.

hamishknight committed:
[Frontend] Add `-exit-on-first-error`

This terminates the compiler as soon as an error
is encountered.

rdar://123630771
